lib/web/sign.go

```diff
@@ -0,0 +1,205 @@
+			ServerNames: []string{req.Hostname},
+			// Include legacy ServerName for compatibility.
+			ServerName:    req.Hostname,
+			TTL:           proto.Duration(req.TTL),
+			RequesterName: proto.DatabaseCertRequest_TCTL,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.TLSCert = resp.Cert
+	key.TrustedCA = []auth.TrustedCerts{{TLSCertificates: resp.CACerts}}
+
+	virtualFS := identityfile.NewInMemoryConfigWriter()
+
+	filesWritten, err := identityfile.Write(identityfile.WriteConfig{
+		OutputPath:           "server",
+		Key:                  key,
+		Format:               req.Format,
+		OverwriteDestination: true,
+		Writer:               virtualFS,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	archiveName := fmt.Sprintf("teleport_mTLS_%s.tar.gz", req.Hostname)
+
+	archiveBytes, err := archiveFromFiles(filesWritten, virtualFS)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Set file name
+	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment;filename="%v"`, archiveName))
+
+	// ServeContent sets the correct headers: Content-Type, Content-Length and Accept-Ranges.
+	// It also handles the Range negotiation
+	http.ServeContent(w, r, archiveName, time.Now(), bytes.NewReader(archiveBytes.Bytes()))
+
+	return nil, nil
+}
+
+// archiveFromFiles builds a Tar Gzip archive in memory, reading the files from the virtual FS
+func archiveFromFiles(files []string, virtualFS identityfile.InMemoryConfigWriter) (*bytes.Buffer, error) {
+	archiveBytes := &bytes.Buffer{}
+
+	gzipWriter := gzip.NewWriter(archiveBytes)
+	defer gzipWriter.Close()
+
+	tarWriter := tar.NewWriter(gzipWriter)
+	defer tarWriter.Close()
+
+	for _, filename := range files {
+		bs, err := virtualFS.Read(filename)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := tarWriter.WriteHeader(&tar.Header{
+			Name: filename,
+			Size: int64(len(bs)),
+
+			// https://www.postgresql.org/docs/current/libpq-ssl.html
+			// On Unix systems, the permissions on the private key file must disallow any access to world or group;
+			//  achieve this by a command such as chmod 0600 ~/.postgresql/postgresql.key.
+			// Alternatively, the file can be owned by root and have group read access (that is, 0640 permissions).
+			Mode: 0600,
+		}); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if _, err := tarWriter.Write(bs); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return archiveBytes, nil
+}
+
+type signCertKeyPairReq struct {
+	Hostname string
+	Format   identityfile.Format
+	TTL      time.Duration
+}
+
+// TODO(marco): only format db is supported
+var supportedFormats = []identityfile.Format{
+	identityfile.FormatDatabase,
+}
+
+func parseSignCertKeyPair(r *http.Request) (*signCertKeyPairReq, error) {
+	reqRaw := struct {
+		Hostname string `json:"hostname,omitempty"`
+		Format   string `json:"format,omitempty"`
+		TTL      string `json:"ttl,omitempty"`
+	}{}
```

Comment:
Why not just define json tags on `signCertKeyPairReq`?

Reply:
I'm using two structs here
One of them contains the fields as received from the JSON bytes and another one contains the actual request

I can use a single struct, loading the JSON into some fields and then `CheckAndSetDefaults` will validate/convert/copy them to the actual request fields

Let me do see how that looks like